### PR TITLE
Add `argparse` and implement `otp` command

### PR DESCRIPTION
**Pull request by robinkrahl**
This patch series introduces the `argparse` crate for argument parsing, adds the `commands` and `options` modules and implements the `otp {set,get,clear,status}` commands.

### CODE REVIEW

**Comment by d-e-s-o** [comment]
Thanks for this pull request, Robin! Unfortunately I may need a few more days until I get to looking at it in more detail.

File: nitrocli/src/options.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        Command::Status => "status",
+      }
+    )
+  }
+}
+
+impl FromStr for Command {
+  type Err = ();
+
+  fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
```
Comment (d-e-s-o) [inline nitrocli/src/options.rs]: It appears we are using:
1) `use std::str::FromStr` (full qualifier in `use` statement)
2) `use std::io` and
3) `std::result::Result` (full qualifier at site of use)

I'd rather we agree on one way to reference symbols. In the past I've typically used the first variant, but it seems uncommon in the community I admit that it may not be the best one. I think 2) is most common. Shall we converge on that?

File: nitrocli/src/commands.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
+// commands.rs
```
Comment (d-e-s-o) [inline nitrocli/src/commands.rs]: It would be great to make the introduction of this file a separate commit. I've actually done that already in [wip/argparse](https://github.com/d-e-s-o/nitrocli/commits/wip/argparse) in order to make reviewing easier. Is that okay with you?

File: nitrocli/src/options.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
+// options.rs
```
Comment (d-e-s-o) [inline nitrocli/src/options.rs]: Would `args.rs` be a more suitable name, given that we work with `argparse`?

File: nitrocli/src/options.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+/// Executes the given parser on the given arguments and handles the result.
+///
+/// This macro executed the given argument parser on the given arguments using the standard output
+/// and error.  It then handles three cases:  If the result is `Ok`, it does nothing as the
+/// arguments have been parsed successfully.  If the result is `Err(0)`, it returns `Ok(())` as the
+/// help message has been printed by `argparse`.  Otherwise it returns `Err(Error::ArgparseError)`
+/// as `argparse` failed.
+macro_rules! parse_args {
```
Comment (d-e-s-o) [inline nitrocli/src/options.rs]: I guess I do understand why you have introduced a macro here, but I think that can be solved with a function and in the process be made more clear. Would it work to simply propagate the error reported by `argparse` up (i.e., extend `Error::ArgparseError` with the error code) and then make the decision as to how to exit in the `run` function?

Largely untested: [argparse.txt](https://github.com/d-e-s-o/nitrocli/files/2704772/argparse.txt)

File: nitrocli/src/options.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    argparse::Store,
+    "the command to execute (clear|close|open|status)",
+  );
+  let _ = parser.refer(&mut args).add_argument(
+    "arguments",
+    argparse::List,
+    "the arguments for the command",
+  );
+  parser.stop_on_first_argument(true);
+  parser.parse_args_or_exit();
```
Comment (d-e-s-o) [inline nitrocli/src/options.rs]: I'd rather not have anything exit the program in some side module. Similarly, I think we should pass in the argument vector to this function such that environment accesses are all located in the main file and dependencies are explicit.

File: nitrocli/src/main.rs [hunk 6]
```diff
@@ -68,176 +68,31 @@
...
-    }
-  }
-}
-
 fn run() -> i32 {
-  let argv: Vec<String> = std::env::args().collect();
-  if argv.len() != 2 {
-    println!("Usage: {} <command>", argv[0]);
-    return 1;
+  let (command, args) = options::parse_arguments();
```
Comment (d-e-s-o) [inline nitrocli/src/main.rs]: Instead of returning the command and `execute`ing it here, do you think it would be clearer to just short-circuit that, rename `parse_arguments` to `handle_arguments` and have it deal with the dispatch of the command?

**Comment by robinkrahl** [comment]
Github seems to butcher e-mail replies, sorry for the mess in the last
comment.

I’ve changed the commits to address most of your points and I removed
the OTP commits.  I did not change the file name and the use directives.
If you want to change them, please let me know (or change the commit).

**Comment by d-e-s-o** [comment]
> Github seems to butcher e-mail replies, sorry for the mess in the last comment.

No worries. I have to say for a platform that is hyped as much as Github is, a couple of the services it provides leave quite a few things to be desired. I find the way reviews are handled absolutely sub par. I can't select multiple lines (or perhaps I am just too incapable to use the UI properly) and reviewing individual patches is quite messy and confusing.

**Comment by robinkrahl** [comment]
> No worries. I have to say for a platform that is hyped as much as Github is, a couple of the services it provides leave quite a few things to be desired. I find the way reviews are handled absolutely sub par. I can't select multiple lines (or perhaps I am just too incapable to use the UI properly) and reviewing individual patches is quite messy and confusing.

… and the issue tracker is very basic too.  It’s fine for small projects like this one, but the stuff they have to do to make it work e. g. for the Rust project is horrific (labels for all kinds of metadata, bots just to manage the issues, etc.).
